refactor aggregates

jingjingwang opened the pull request:
- Unify aggregate operators: remove Single/MultiGroupByAggregate, make StreamingAggregate as a subclass of Aggregate, remove CountAllAggregator
- Introduce an abstraction of tuple hash table, make Aggregate/SymmetricHashJoin use it
- Make aggregators column-oriented: the internal state contains columns of "basic" aggregation ops (e.g. SUM, COUNT), and "complex" aggregation ops which emit results are mapped to them (e.g. AVG). Only internal aggregations are applied to input tuples. Results are generated by emit aggregations.

Works with https://github.com/uwescience/raco/pull/553.

senderista commented:
Could you find a way to avoid re-opening a PR that's under review? I have to copy/paste a few dozen comments by hand into the new PR...

senderista commented on src/edu/washington/escience/myria/storage/Tuple.java:
```diff
@@ -30,16 +27,26 @@
    * @param schema the schema of the tuple
    */
   public Tuple(final Schema schema) {
+    data = new MutableTupleBuffer(schema);
```
What is the rationale for replacing a simple `ArrayList<Field>` with `MutableTupleBuffer`? It appears that `MutableTupleBuffer` carries a lot of overhead for a single-row table, especially when we're creating lots of `Tuple`s (e.g. in your implementation of `Aggregator.addRow()`). Actually, I wonder if we need two versions of `Tuple`: an immutable one which is simply a `ReadableTable` view over selected columns in a selected row of a `ReadableTable`, and a mutable one which implements `AppendableTable` but holds only one row. It seems that combining these two responsibilities in a single class creates unnecessary overhead.

senderista commented on src/edu/washington/escience/myria/util/MyriaArrayUtils.java:
```diff
@@ -153,4 +153,19 @@ private MyriaArrayUtils() {}
+   * @param length the length of the array.
+   * @return an array of the numbers [start, start+length).
+   */
+  public static int[] range(final int start, final int length) {
```
We're on Java 8, so: `IntStream.range(start, start+length).toArray()`

senderista commented on test/edu/washington/escience/myria/operator/AggregateTest.java:
```diff
@@ -144,10 +139,8 @@ private TupleBatch doAggOpsToCol(
-   *
-   * This variant uses a SingleGroupByAggregate in order to do extra testing of the Aggregators by hitting a function
-   * that Aggregate does not use.
+   * expected. This variant uses a SingleGroupByAggregate in order to do extra testing of the Aggregators by hitting a
```
change reference to removed class `SingleGroupByAggregate`

jingjingwang commented on src/edu/washington/escience/myria/api/encoding/OperatorEncoding.java:
```diff
@@ -14,12 +14,11 @@
  * A JSON-able wrapper for the expected wire message for an operator. To add a new operator, two things need to be done.
- * 1. Create an Encoding class that extends OperatorEncoding.
- * 2. Add the operator to the list of (alphabetically sorted) JsonSubTypes below.
+ * 1. Create an Encoding class that extends OperatorEncoding. 2. Add the operator to the list of (alphabetically sorted)
```
I wish there is a way to tell Eclipse not to reformat it... :)

jingjingwang commented:
For `TupleHashTable`, actually I'm expecting a lot of collisions from joins, since one join key may correspond to many tuples. Aggregates are fine. Since usually joins have much larger hash tables, I was not very motivated to do the optimization. I agree that it is a cool design if collisions are rare.

jingjingwang commented on src/edu/washington/escience/myria/api/encoding/OperatorEncoding.java:
```diff
@@ -50,7 +49,7 @@
   @Type(name = "Merge", value = MergeEncoding.class),
   @Type(name = "MergeJoin", value = MergeJoinEncoding.class),
-  @Type(name = "MultiGroupByAggregate", value = MultiGroupByAggregateEncoding.class),
+  @Type(name = "Aggregate", value = AggregateEncoding.class),
```
I wish there is a way to tell Eclipse to reformat it... :)

jingjingwang commented on src/edu/washington/escience/myria/operator/Apply.java:
```diff
@@ -128,7 +127,7 @@ protected TupleBatch fetchNextReady() throws DbException, InvocationTargetExcept
         if (onlySingleValuedExpressions()) {
           outputBuffer.absorb(
-              new TupleBatch(getSchema(), resultColumnsForTB, inputTuples.numTuples()));
+              new TupleBatch(getSchema(), resultColumnsForTB, inputTuples.numTuples()), true);
```
This `true` is not for `new TupleBatch()`, it is for `outputBuffer.absorb()`.

jingjingwang commented on src/edu/washington/escience/myria/operator/TupleHashTable.java:
```diff
@@ -0,0 +1,131 @@
+  private static final long serialVersionUID = 1L;
+
+  /** Map from hash codes to indices. */
+  private transient IntObjectHashMap<IntArrayList> keyToIndices;
```
Exactly.

jingjingwang commented on src/edu/washington/escience/myria/operator/SymmetricHashJoin.java:
```diff
@@ -32,279 +27,93 @@
-   * The column indices for comparing of child 1.
-   */
+  /** The column indices for comparing of the left child. */
   private final int[] leftCompareIndx;
```
After another though I think `leftCompareColumns` sounds better, it also matches `leftAnswerColumns`.

jingjingwang commented on src/edu/washington/escience/myria/operator/agg/Aggregate.java:
```diff
@@ -1,113 +1,155 @@
+      for (Aggregator agg : emitAggs) {
+        columns.addAll(agg.emitOutput(tb));
+      }
+      resultBuffer.absorb(new TupleBatch(getSchema(), columns), true);
```
Same as above, this `true` is for `absorb()`.

jingjingwang commented on src/edu/washington/escience/myria/operator/agg/BooleanAggregator.java:
```diff
@@ -1,112 +1,78 @@
-      b.count = LongMath.checkedAdd(b.count, 1);
+    ReplaceableColumn toCol = to.getColumn(stateCols[0], toRow);
+    final int inColumRow = to.getInColumnIndex(toRow);
+    switch (aggOp) {
```
I'm not trying to add any new ops here, just refactoring.

jingjingwang commented on src/edu/washington/escience/myria/operator/agg/AggregatorFactory.java:
```diff
@@ -16,18 +15,44 @@
+   * @return aggregators aggregators for internal state.
+   * @throws DbException if there is an error creating the aggregators
+   */
+  List<Aggregator> generateInternalAggs(final Schema inputSchema, final int offset)
```
Do you think using expressions like in `Apply` would be a good idea? Actually `Aggregate` reminds me of `Apply` a lot especially after this refactoring.

senderista commented:
Right, that's why I suggested that maybe we should have different interfaces/impls for `TupleHashTable` based on whether keys are expected to reference unique indices. For aggregates this will be true, for joins no. I don't think we should necessarily assume that aggregate groups will be coarse-grained; someone will eventually run a query where that is not the case. Anyway, I opened https://github.com/uwescience/myria/issues/876 for this issue, so we can drop it from the PR :)

jingjingwang commented on src/edu/washington/escience/myria/operator/agg/StringAggregator.java:
```diff
@@ -1,141 +1,108 @@
-    super(aFieldName, aggOps);
-    fromColumn = column;
+  protected StringAggregator(
+      final String inputName, final int column, final AggregationOp aggOp, final int[] stateCols) {
```
Because an Aggregator only performs one op. The factory will generates multiple aggregators if we have multiple ops on one column.

senderista commented on src/edu/washington/escience/myria/operator/agg/UserDefinedAggregatorFactory.java:
```diff
@@ -105,13 +106,12 @@ public UserDefinedAggregatorFactory(
-      evaluator.compile();
-      emitEvaluators.add(evaluator);
-    }
+    // emitEvaluators = new ArrayList<>(emitters.size());
```
This can all be removed, right?

senderista commented on src/edu/washington/escience/myria/operator/agg/UserDefinedAggregatorFactory.java:
```diff
@@ -123,8 +123,7 @@ public UserDefinedAggregatorFactory(
     return ImmutableList.of(
-        new UserDefinedAggregator(
-            initEvaluator, updateEvaluator, emitEvaluators, resultSchema, stateSchema, offset));
+        new UserDefinedAggregator(initEvaluator, updateEvaluator, resultSchema, stateSchema));
```
OK, so the `emitEvaluators` are now computed from `emitters` in the `Aggregate` operator?

jingjingwang replied:
Yep, so Aggregators are not responsible for emitting results anymore. They only need to aggregate.

senderista commented on src/edu/washington/escience/myria/operator/agg/UserDefinedAggregatorFactory.java:
```diff
@@ -157,16 +156,35 @@ private ScriptEvalInterface getEvalScript(
+            .append(varCount)
+            .append(");\n");
+      } else {
+        // state.replaceType(I, stateRow, valI);
```
?

senderista commented on src/edu/washington/escience/myria/operator/agg/UserDefinedAggregatorFactory.java:
```diff
@@ -157,16 +156,35 @@ private ScriptEvalInterface getEvalScript(
-          .append(varCount)
-          .append(");\n");
+      if (param.getStateSchema() == null) {
+        // state.putType(I, valI);
```
?

jingjingwang replied:
It's to show what expression is being constructed. I simply followed the format of the previous comment.

senderista commented:
Unfortunately most of my comments on your last refactoring commit are hidden behind "Show outdated"...

jingjingwang commented on src/edu/washington/escience/myria/operator/agg/StringAggregator.java:
```diff
@@ -27,7 +27,7 @@ protected StringAggregator(
 
   /** Placeholder as MIN/MAX value of String. */
-  private static final String STRING_INIT_VALUE = "STRING_INIT_VALUE";
+  private static final String STRING_INIT_VALUE = "NULL";
```
I have to remove the null check in `StringColumnBuilder` then.

senderista replied:
How about we use the Unicode BOM instead then:
```
private static final String STRING_INIT_VALUE = "\uFEFF";
```

jingjingwang commented on src/edu/washington/escience/myria/expression/evaluate/GenericEvaluator.java:
```diff
@@ -216,7 +217,7 @@ public EvaluatorResult evaluateColumn(final TupleBatch tb) throws InvocationTarg
     final WritableColumn resultsWriter = resultsBuffer.asWritableColumn(0);
     for (int rowIdx = 0; rowIdx < tb.numTuples(); ++rowIdx) {
-      eval(tb, rowIdx, countsWriter, resultsWriter, null);
+      eval(tb, rowIdx, tb, rowIdx, resultsWriter, countsWriter);
```
This is somewhat a hack since this `tb` could be either the input tuple batch from child or a slice of the `TupleHashTable` state. However the stateful emit expressions only know to fetch data from the "state" even if the "state" serves more like an input tuple batch in this case. I added a comment and will think about how to make it more explicit.

jingjingwang commented on src/edu/washington/escience/myria/operator/agg/Aggregate.java:
```diff
@@ -138,20 +160,27 @@ protected void init(final ImmutableMap<String, Object> execEnvVars) throws DbExc
+        GenericEvaluator evaluator =
+            new GenericEvaluator(
+                exp, new ExpressionOperatorParameter(newStateSchema, newStateSchema));
+        // if (evaluator.needsCompiling()) {
```
I removed everything about `needsCompiling()` in the new commit. First, it doesn't hurt to always compile since it only happens in `init()`, and second, now we need to evaluate tuple-by-tuple when the input and the output tuple batches have different sizes, so we need the compiled expression. Found some old discussions in https://github.com/uwescience/myria/pull/548.

jingjingwang commented on src/edu/washington/escience/myria/operator/agg/BooleanAggregator.java:
```diff
@@ -17,19 +13,22 @@
+    ReplaceableColumn toCol = to.getColumn(offset, toRow);
     final int inColumRow = to.getInColumnIndex(toRow);
     switch (aggOp) {
       case COUNT:
```
I don't plan to add them in this PR.

senderista replied:
tracking this in https://github.com/uwescience/myria/issues/880

jingjingwang commented on src/edu/washington/escience/myria/operator/agg/LongAggregator.java:
```diff
@@ -63,42 +61,6 @@ public void addRow(
-  }
-
-  @Override
   protected boolean isSupported(final AggregationOp aggOp) {
```
Here the "supported" means the exposed aggregate such as AVG but not internal aggregates from 'getInternalOps()'. `addRow()` is only called with an `aggOp` from 'getInternalOps()' . Do you think we should remove the check? It's more like a check for RACO. I feel that it doesn't hurt to double-check here.

jingjingwang commented on src/edu/washington/escience/myria/operator/agg/PrimitiveAggregatorFactory.java:
```diff
@@ -68,42 +74,68 @@ public PrimitiveAggregatorFactory(final Integer column, final AggregationOp aggO
+    for (int i = 0; i < aggOps.length; ++i) {
+      String name = aggOps[i].toString().toLowerCase() + "_" + inputSchema.getColumnName(column);
+      switch (aggOps[i]) {
+        case COUNT:
```
I'm not sure if `getAggColumnType()` has what you want.

jingjingwang commented on src/edu/washington/escience/myria/expression/evaluate/ExpressionEvalInterface.java:
```diff
@@ -15,16 +15,18 @@
    * {@link edu.washington.escience.myria.operator.StatefulApply}. The variables will be fetched from the tuple buffer
    * using the rowId provided in {@link edu.washington.escience.myria.expression.VariableExpression}.
```
It's an outdated comment, will remove in the new commit.

jingjingwang commented on src/edu/washington/escience/myria/operator/agg/Aggregate.java:
```diff
@@ -103,14 +110,29 @@ protected TupleBatch fetchNextReady() throws DbException {
+        TupleBatch state =
+            tb.selectColumns(MyriaArrayUtils.range(stateOffset, stateOffset + stateSize));
+        for (GenericEvaluator eval : emitEvals.subList(emitOffset, emitOffset + emitSize)) {
+          columns.addAll(eval.evaluateColumn(state).getResultColumns());
```
Yep I also have realized it. I don't know if we can assume that `Aggregate` would only have singe-valued emit expressions? For now it seems to be true, even when Python UDFs are involved. Will check with @parmitam .

jingjingwang commented:
I have addressed all comments and also adopted new changes in master. (Is there a way to make coverall pass? I didn't change much...) There are several things that I need to confirm with @parmitam, but the new commit is good to be reviewed first.

jingjingwang commented:
I will do #876 and #878 in other PRs.

jingjingwang commented:
I also cherry-picked two commits from #867 here since one is relevant.

jingjingwang commented on test/edu/washington/escience/myria/operator/UserDefinedAggregatorTest.java:
```diff
@@ -197,7 +197,7 @@ public void testRowOfMax() throws Exception {
         assertEquals(Type.LONG_TYPE, result.getSchema().getColumnType(0));
         assertEquals(Type.STRING_TYPE, result.getSchema().getColumnType(1));
-        assertEquals(9999, result.getLong(0, 0));
+        assertEquals(10000, result.getLong(0, 0));
```
I think the old test was wrong since the first column is the row index, and the expression is `+1`, so the last row should have row index to be 10000. I didn't check why the old test was wrong though.

jingjingwang commented on src/edu/washington/escience/myria/expression/evaluate/GenericEvaluator.java:
```diff
@@ -108,14 +129,36 @@ public void compile() throws DbException {
    */
   public void eval(
-      @Nonnull final ReadableTable input,
+      @Nullable final ReadableTable input,
```
When an emit expression is evaluated, the `input` is null but the `state` is not, although the `state` works as an `input`.

jingjingwang commented on src/edu/washington/escience/myria/expression/evaluate/GenericEvaluator.java:
```diff
@@ -65,38 +78,46 @@ public GenericEvaluator(
-                    Expression.RESULT,
-                    Expression.COUNT
-                  });
+      if (script.contains("append")) {
```
I agree, I think we might be able to clean it when solving #878.

jingjingwang commented on src/edu/washington/escience/myria/expression/evaluate/ExpressionEvalAppendInterface.java:
```diff
@@ -0,0 +1,29 @@
+   * @param count a column storing the number of results returned from this row
+   */
+  void evaluate(
+      final ReadableTable input,
```
It's actually `@Nullable` for emit expressions.

jingjingwang commented on src/edu/washington/escience/myria/expression/evaluate/PythonUDFEvaluator.java:
```diff
@@ -60,21 +52,20 @@
-  /** tuple size to be sent to the python process, this is equal to the number of children of the expression. */
-  private int numColumns = -1;
-
+  private Set<Integer> stateColumns;
```
It's simply for making `contains()` faster (in theory).

jingjingwang commented on src/edu/washington/escience/myria/expression/evaluate/PythonUDFEvaluator.java:
```diff
@@ -143,77 +128,70 @@ public void compile() {
+      ColumnBuilder<?> output = ColumnFactory.allocateColumn(outputType);
+      /* TODO: Leaving the count column to be null for now since since it's not used by Python evaluator for aggregate.
+       * A better design is to let the Aggregator emit two columns or even multiple columns. */
+      readFromStream(null, output);
```
Then we need to write to the `WriteableColumn` in `PythonUDFEvaluator.eval()`.

jingjingwang commented on src/edu/washington/escience/myria/operator/Apply.java:
```diff
@@ -293,9 +293,7 @@ protected void init(final ImmutableMap<String, Object> execEnvVars) throws DbExc
-      if (evaluator.needsCompiling()) {
-        evaluator.compile();
-      }
+      evaluator.compile();
```
Yep! It is finally reasonable.

jingjingwang commented on src/edu/washington/escience/myria/expression/evaluate/PythonUDFEvaluator.java:
```diff
@@ -60,21 +52,20 @@
   /** is expression a flatmap? */
   private Boolean isMultiValued = false;
+  /** Tuple buffers for each group key. */
+  private IntObjectHashMap<TupleBatchBuffer> buffer;
```
It's a little bit awkward, since each entry in this `buffer` was designed to match a row of the `state` in `Aggregate`.  If we change it to a `TupleHashTable`, then we need to pass the values of keys of each row in `state` to this `TupleHashTable`, hash them, and check if the keys indeed match, in order to send all tuples of a group to Python. I think there is some CPU overhead. One the other hand, the memory overhead of this design comparing to using `TupleHashTable`, is the Java object overhead of a list of Columns per group key, which may not be very bad to me. I agree that it would be the best to get rid of this memory overhead without sacrificing time complexity.

senderista commented on src/edu/washington/escience/myria/expression/evaluate/ExpressionEvalAppendInterface.java:
```diff
@@ -14,16 +17,16 @@
    *
    * @param input the input tuple batch
```
If this is optional (`@Nullable`) then it should be documented as such.

senderista commented on test/edu/washington/escience/myria/operator/AggregateTest.java:
```diff
@@ -49,17 +49,9 @@
-   * COUNT is always long
-   *
-   * AVG and STDEV are always double
+   * Ensure that the given Schema matches the expected numeric aggregate types for the given Type. All numeric aggs, in
```
This is messed up again.

jingjingwang commented:
I think coverall was messed up since there is no way to change the coverage by 0.8% through modifying a few lines of Java docs...

senderista commented on src/edu/washington/escience/myria/storage/TupleBuffer.java:
```diff
@@ -349,4 +349,18 @@ public ReadableColumn asColumn(final int column) {
+   * @param sourceColumn the column from which data will be retrieved.
+   * @param sourceRow the row in the source column from which data will be retrieved.
+   */
+  public final void put(
```
Long overdue!

jingjingwang replied:
Yeah! I can't believe this interface was not there.
